# Домашнее задание 4 - предобработка признаков, pandas

Дата выдачи: 28.09.2022

Срок сдачи: 09.10.2022 23:59MSK

### О задании

Практическое задание 1 посвящено изучению основных библиотек для анализа данных, а также линейных моделей и методов их обучения. Вы научитесь:
 * применять библиотеки NumPy и Pandas для осуществления желаемых преобразований;
 * подготавливать данные для обучения линейных моделей;
 * обучать линейную, Lasso и Ridge-регрессии при помощи модуля scikit-learn;
 * реализовывать обычный и стохастический градиентные спуски;
 * обучать линейную регрессию для произвольного функционала качества.
 

### Оценивание и штрафы

Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов. Кроме того, некоторые из заданий являются опциональными (необязательными), однако за их выполнение можно получить дополнительные баллы, которые позднее будут учитываться при проставлении оценок автоматом по курсу.

Сдавать задание после указанного срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце Вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник). 

Неэффективная реализация кода может негативно отразиться на оценке.


### Формат сдачи
Для сдачи задания переименуйте получившийся файл \*.ipynb в соответствии со следующим форматом: *HW4_Username.ipynb*, где *Username* — Ваша фамилия и инициалы на латинице (например, *HW4_IvanovII.ipynb*). Необходимо выложить файл с решением в свой репозиторий github.

## Библиотеки для анализа данных

### NumPy

Во всех заданиях данного раздела запрещено использовать циклы  и list comprehensions. Под вектором и матрицей в данных заданиях понимается одномерный и двумерный numpy.array соответственно.

In [1]:
import numpy as np

**1. (0.2 балла)** Реализуйте функцию, возвращающую максимальный элемент в векторе x среди элементов, перед которыми стоит нулевой. Для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответом является 5. Если нулевых элементов нет, функция должна возвращать None.


In [2]:
def max_element(arr):
    
    if arr[0] != 0:
        arr = arr[1:]

    nonzeros_indices = np.nonzero(arr)[0]

    if len(nonzeros_indices) == len(arr):
        return None
    
    nonzero_before_zero = arr[nonzeros_indices - 1] == 0

    return arr[nonzeros_indices[nonzero_before_zero]].max()

In [4]:
x = np.array([8, 1, 2, 6, 0, 5, 0, 9, 2])

max_element(x)

9

**2. (0.2 балла)** Реализуйте функцию, принимающую на вход матрицу и некоторое число и возвращающую ближайший к числу элемент матрицы. Например: для X = np.arange(0,10).reshape((2, 5)) и v = 3.6 ответом будет 4.

In [5]:
def nearest_value(X, v):
    X = X.astype("float")
    X_dst = np.abs(np.abs(X) - v)
    min_dst = np.argmin(X_dst)
    return X[min_dst // X.shape[1], min_dst % X.shape[1]]

In [7]:
X = np.arange(0,10).reshape((5, 2))
v = 3.3

nearest_value(X, v)

3.0

**3. (0.2 балла)** Реализуйте функцию scale(X), которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает выборочное среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на случайной матрице (для её генерации можно использовать, например, функцию [numpy.random.randint](http://docs.scipy.org/doc/numpy/reference/generated/numpy.random.randint.html)).

In [8]:
def scale(X):
    
    s_d = X.std(axis=0)
    s_d[s_d == 0] = 1

    X = X - X.mean(axis=0)
    X /= s_d
    
    return X

In [9]:
A = np.random.randint(-5, 16, (3, 5))
A

array([[14, -1, -5, -1, 13],
       [ 5, -3, -1, 12, 11],
       [-2, -1,  5,  2,  6]])

In [10]:
scale(A)

array([[ 1.2724665 ,  0.70710678, -1.13554995, -0.95961623,  1.01904933],
       [-0.10179732, -1.41421356, -0.16222142,  1.37944833,  0.33968311],
       [-1.17066918,  0.70710678,  1.29777137, -0.4198321 , -1.35873244]])

**4. (0.2 балла)** Реализуйте функцию, которая для заданной матрицы находит:
 - определитель
 - след
 - наименьший и наибольший элементы
 - норму Фробениуса
 - собственные числа
 - обратную матрицу

Для тестирования сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}$(10,1)

In [11]:
def get_stats(X):
    
    print("Определитель: ", np.linalg.det(X))
    print("След: ", np.trace(X))
    print("Наибольший элемент: ", X.max())
    print("Наименьший элемент: ", X.min())
    print("Норма Фробениуса: ", np.linalg.norm(X, ord='fro'))
    print("Собственные числа: ", np.linalg.eig(X)[0])
    print("Обратная матрица: ", '\n', np.linalg.inv(X))
    
    return

In [12]:
X = np.random.normal(10, 1, (4, 4))
get_stats(X)

Определитель:  -13.867654017749881
След:  41.06156341218518
Наибольший элемент:  10.990974549434895
Наименьший элемент:  7.64217634736665
Норма Фробениуса:  39.97028120855574
Собственные числа:  [39.72839666  1.47727718 -0.56346037  0.41934995]
Обратная матрица:  
 [[-0.26085789 -1.53384848 -1.68599503  3.41281078]
 [ 0.12631579  2.0830454   0.84289144 -2.87510506]
 [-0.27213734 -0.45776075  0.28815529  0.41320567]
 [ 0.40734875 -0.19752271  0.59644921 -0.79835535]]


**5. (0.2 балла)** Повторите 100 раз следующий эксперимент: сгенерируйте две матрицы размера 10×10 из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

In [14]:
max_elem = np.array([])
for exp_num in range(100):
    C1, C2 = np.random.normal(10, 1, (10, 10)), np.random.normal(10, 1, (10, 10))
    C1_C2 = C1.dot(C2)
    max_elem = np.append(max_elem, np.max(C1_C2))

print("Cреднее значение по экспериментам у максимальных элементов:", np.mean(max_elem))
print('95-процентная квантиль:', np.quantile(max_elem, 0.95)) 

Cреднее значение по экспериментам у максимальных элементов: 1099.2147048698776
95-процентная квантиль: 1151.3527470050217


### Pandas

![](https://metrouk2.files.wordpress.com/2015/10/panda.jpg)

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

[Данные](https://www.dropbox.com/s/dvfitn93obn0rql/2008.csv?dl=0) и их [описание](http://stat-computing.org/dataexpo/2009/the-data.html)

Airline on-time performance

Have you ever been stuck in an airport because your flight was delayed or cancelled and wondered if you could have predicted it if you'd had more data? This is your chance to find out.

The data
The data set is available for download here.
The data consists of flight arrival and departure details for all commercial flights within the USA, from October 1987 to April 2008. This is a large dataset: there are nearly 120 million records in total, and takes up 1.6 gigabytes of space compressed and 12 gigabytes when uncompressed.

Understanding and preparing the data
In order to answer above questions, we are going to analyze the provided dataset, containing up to 1936758 ### different internal flights in the US for 2008 and their causes for delay, diversion and cancellation

The data comes from the U.S. Department of Transportation’s (DOT) Bureau of Transportation Statistics (BTS). Meta data explanations

This dataset is composed by the following variables:

**Year** 2008 **Month** 1-12 **DayofMonth** 1-31 **DayOfWeek** 1 (Monday) - 7 (Sunday)  
**DepTime** actual departure time (local, hhmm)  
**CRSDepTime** scheduled departure time (local, hhmm)  
**ArrTime** actual arrival time (local, hhmm)  
**CRSArrTime** scheduled arrival time (local, hhmm)  
**UniqueCarrier** unique carrier code  
**FlightNum** flight number  
**TailNum** plane tail number: aircraft registration, unique aircraft identifier  
**ActualElapsedTime** in minutes  
**CRSElapsedTime** in minutes  
**AirTime** in minutes  
**ArrDelay** arrival delay, in minutes: A flight is counted as “on time” if it operated less than 15 minutes later the scheduled time shown in the carriers’ Computerized Reservations Systems (CRS).  
**DepDelay** departure delay, in minutes  
**Origin** origin IATA airport code  
**Dest** destination IATA airport code  
**Distance** in miles  
**TaxiIn** taxi in time, in minutes  
**TaxiOut** taxi out time in minutes  
**Cancelled** *was the flight cancelled  
**CancellationCode** reason for cancellation (A = carrier, B = weather, C = NAS, D = security)  
**Diverted** 1 = yes, 0 = no  
**CarrierDelay** in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.  
**WeatherDelay** in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.  
**NASDelay** in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.  
**SecurityDelay** in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.  
**LateAircraftDelay** in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagation.

In [15]:
import pandas as pd
%matplotlib inline

In [16]:
db = pd.read_csv('2008.csv')

**6. (0.3 балла)** Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)

In [17]:
db['CancellationCode'].value_counts().idxmax()

'A'

**7. (0.3 балла)** Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.

In [18]:
print('Среднее пройденное расстояние:', db['Distance'].mean())
print('Минимальное пройденное расстояние:', db['Distance'].min())
print('Максимальное пройденное расстояние:', db['Distance'].max())

Среднее пройденное расстояние: 724.5082571428571
Минимальное пройденное расстояние: 31
Максимальное пройденное расстояние: 4962


**8. (0.3 балла)** Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?

In [19]:
db[db['Distance'] == db['Distance'].min()][['Year', 'Month', 'DayofMonth', 'FlightNum', 'Dest', 'Distance']].sort_values(['Year', 'Month', 'DayofMonth'], axis=0)

Year  Month  DayofMonth  FlightNum Dest  Distance
48112  2008      2          28         64  WRG        31
27534  2008      3          11         64  WRG        31
46082  2008      8           9         65  PSG        31
17349  2008      8          18         64  WRG        31
6958   2008     12          26         65  PSG        31
1116   2008     12          30         65  PSG        31

In [20]:
db[(db['FlightNum'] == 64) | (db['FlightNum'] == 65)]['Distance'].value_counts()

1747    12
82      11
1005     9
123      6
31       6
571      6
680      5
533      4
372      4
581      4
2454     3
2381     2
682      1
414      1
359      1
281      1
883      1
386      1
Name: Distance, dtype: int64

**9. (0.3 балла)** Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?

In [21]:
db['Origin'].value_counts().idxmax()

'ATL'

**10. (0.3 балла)** Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?

In [22]:
db.groupby('Origin')['AirTime'].mean()

Origin
ABE    88.266667
ABI    36.400000
ABQ    93.454321
ABY    35.714286
ACK    50.800000
         ...    
WRG    18.000000
XNA    85.945736
YAK    35.900000
YKM    79.000000
YUM    47.470588
Name: AirTime, Length: 297, dtype: float64

In [23]:
db.groupby('Origin')['AirTime'].mean().idxmax()

'SJU'

**11. (0.5 балла)** Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [24]:
all_air = db.groupby('Origin').filter(lambda x: len(x) >= 1000).groupby('Origin').size()

all_air_delay = db.groupby('Origin').filter(lambda x: len(x) >= 1000)[db['DepDelay'] > 0].groupby('Origin').size()

p = all_air_delay / all_air
p.idxmax()

<ipython-input-24-860672a42bf1>:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  all_air_delay = db.groupby('Origin').filter(lambda x: len(x) >= 1000)[db['DepDelay'] > 0].groupby('Origin').size()


'EWR'

## Линейная регрессия

В этой части мы разберемся с линейной регрессией, способами её обучения и измерением качества ее прогнозов. 

Будем рассматривать датасет из предыдущей части задания для предсказания времени задержки отправления рейса в минутах (DepDelay). Отметим, что под задержкой подразумевается не только опоздание рейса относительно планируемого времени вылета, но и отправление до планируемого времени.

### Подготовка данных

**12. (0.5 балла)** Считайте выборку из файла при помощи функции pd.read_csv и ответьте на следующие вопросы:
   - Имеются ли в данных пропущенные значения?
   - Сколько всего пропущенных элементов в таблице "объект-признак"?
   - Сколько объектов имеют хотя бы один пропуск?
   - Сколько признаков имеют хотя бы одно пропущенное значение?

In [25]:
db.shape

(70000, 29)

In [26]:
db.isnull().any()

Year                 False
Month                False
DayofMonth           False
DayOfWeek            False
DepTime               True
CRSDepTime           False
ArrTime               True
CRSArrTime           False
UniqueCarrier        False
FlightNum            False
TailNum               True
ActualElapsedTime     True
CRSElapsedTime        True
AirTime               True
ArrDelay              True
DepDelay              True
Origin               False
Dest                 False
Distance             False
TaxiIn                True
TaxiOut               True
Cancelled            False
CancellationCode      True
Diverted             False
CarrierDelay          True
WeatherDelay          True
NASDelay              True
SecurityDelay         True
LateAircraftDelay     True
dtype: bool

In [27]:
print('Всего пропущенных элементов:', db.isnull().sum().sum())
print('Кол-во объектов, имеющих хотя бы один пропуск:', db.isnull().any(axis=1).sum())
print('Кол-во признаков, имеющих хотя бы один пропуск:', db.isnull().any(axis=0).sum())

Всего пропущенных элементов: 355215
Кол-во объектов, имеющих хотя бы один пропуск: 70000
Кол-во признаков, имеющих хотя бы один пропуск: 16


Как вы понимаете, также не имеет смысла рассматривать при решении поставленной задачи объекты с пропущенным значением целевой переменной. В связи с этим ответьте на следующие вопросы и выполните соответствующие действия:
- Имеются ли пропущенные значения в целевой переменной? Да
- Проанализируйте объекты с пропущенными значениями целевой переменной. Чем вызвано это явление? Что их объединяет? Можно ли в связи с этим, на ваш взгляд, исключить какие-то признаки из рассмотрения? Обоснуйте свою точку зрения.

Исключите из выборки объекты **с пропущенным значением целевой переменной и со значением целевой переменной, равным 0**, а также при необходимости исключите признаки в соответствии с вашим ответом на последний вопрос из списка и выделите целевую переменную в отдельный вектор, исключив её из матрицы "объект-признак".

In [36]:
db[db['DepDelay'].isnull()].head(15)

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
11   2008      3           6          4      NaN        1605      NaN   
78   2008      1          19          6      NaN         750      NaN   
103  2008      7          24          4      NaN        1230      NaN   
112  2008      2           4          1      NaN         620      NaN   
128  2008      1          21          1      NaN        1838      NaN   
194  2008      1          28          1      NaN        1640      NaN   
204  2008      6          13          5      NaN        1229      NaN   
236  2008      7          19          6      NaN         544      NaN   
257  2008      4          19          6      NaN        1345      NaN   
267  2008      9          23          2      NaN        1400      NaN   
269  2008      2          17          7      NaN        1345      NaN   
288  2008      9          14          7      NaN        1115      NaN   
352  2008      1          20          7      NaN        2023      NaN   
375  2008     12          22          1      NaN         935      NaN   
423  2008      1          23          3      NaN        1640      NaN   

     CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
11         1940            AA        484  ...    NaN      NaN          1   
78          930            EV       4772  ...    NaN      NaN          1   
103        1355            FL        816  ...    NaN      NaN          1   
112         758            NW       1281  ...    NaN      NaN          1   
128        2005            UA       1251  ...    NaN      NaN          1   
194        1719            UA        753  ...    NaN      NaN          1   
204        1421            OH       5765  ...    NaN      NaN          1   
236         917            OO       6151  ...    NaN      NaN          1   
257        1435            AA       1853  ...    NaN      NaN          1   
267        1515            WN       2049  ...    NaN      NaN          1   
269        1535            MQ       3677  ...    NaN      NaN          1   
288        1222            US        535  ...    NaN      NaN          1   
352        2122            AS        190  ...    NaN      NaN          1   
375        1208            DL       1053  ...    NaN      NaN          1   
423        1730            MQ       3965  ...    NaN      NaN          1   

     CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
11                  B         0           NaN          NaN      NaN   
78                  B         0           NaN          NaN      NaN   
103                 A         0           NaN          NaN      NaN   
112                 B         0           NaN          NaN      NaN   
128                 A         0           NaN          NaN      NaN   
194                 C         0           NaN          NaN      NaN   
204                 A         0           NaN          NaN      NaN   
236                 A         0           NaN          NaN      NaN   
257                 A         0           NaN          NaN      NaN   
267                 A         0           NaN          NaN      NaN   
269                 B         0           NaN          NaN      NaN   
288                 A         0           NaN          NaN      NaN   
352                 A         0           NaN          NaN      NaN   
375                 A         0           NaN          NaN      NaN   
423                 C         0           NaN          NaN      NaN   

     SecurityDelay  LateAircraftDelay  
11             NaN                NaN  
78             NaN                NaN  
103            NaN                NaN  
112            NaN                NaN  
128            NaN                NaN  
194            NaN                NaN  
204            NaN                NaN  
236            NaN                NaN  
257            NaN                NaN  
267            NaN                NaN  
269            NaN                NaN  
28

Кол-во пропущенных элементов в признаке DepDelay:

In [37]:
db['DepDelay'].isnull().sum()

1399

Сумма пропусков по признакам:

In [38]:
db[db.isnull()['DepDelay']].isnull().sum()

Year                    0
Month                   0
DayofMonth              0
DayOfWeek               0
DepTime              1399
CRSDepTime              0
ArrTime              1399
CRSArrTime              0
UniqueCarrier           0
FlightNum               0
TailNum               820
ActualElapsedTime    1399
CRSElapsedTime          4
AirTime              1399
ArrDelay             1399
DepDelay             1399
Origin                  0
Dest                    0
Distance                0
TaxiIn               1399
TaxiOut              1399
Cancelled               0
CancellationCode        0
Diverted                0
CarrierDelay         1399
WeatherDelay         1399
NASDelay             1399
SecurityDelay        1399
LateAircraftDelay    1399
dtype: int64

Во всех элементах, где нет информации про DepDelay, также нет никакой другой информации про время.

Кол-во отмененных рейсов среди тех, где нет инф. про DepDelay:

In [39]:
db[db.isnull()['DepDelay']]['Cancelled'].sum()

1399

Отмененные рейсы, у которых ЕСТЬ информация по DepDelay:

In [40]:
db[(db['Cancelled'] == 1) & (db['DepDelay'].isnull() == False)]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
5976   2008     11          29          6   1325.0        1301      NaN   
15680  2008     12          19          5    954.0         935      NaN   
16382  2008     12           6          6   1806.0        1728      NaN   
20622  2008     10           7          2    844.0         845      NaN   
22462  2008     12          17          3   2132.0        2134      NaN   
30057  2008     12           9          2   1358.0        1330      NaN   
33611  2008     10          11          6    826.0         830      NaN   
37581  2008     12           3          3   2234.0        2235      NaN   
40343  2008     11          10          1   1440.0        1445      NaN   
50401  2008     11          29          6   2232.0        2234      NaN   
50810  2008     11          19          3   1559.0        1606      NaN   
67078  2008     12           5          5   1517.0        1510      NaN   

       CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
5976         1354            EV       5401  ...    NaN      NaN          1   
15680        1252            XE       2573  ...    NaN      NaN          1   
16382        1827            OO       5790  ...    NaN     11.0          1   
20622        1039            YV       7302  ...    NaN      NaN          1   
22462        2232            YV       2859  ...    NaN     28.0          1   
30057        1700            OO       2874  ...    NaN      NaN          1   
33611        1015            OO       3721  ...    NaN      NaN          1   
37581        2318            OO       5426  ...    NaN     16.0          1   
40343        1731            EV       5309  ...    NaN      NaN          1   
50401          11            OO       6412  ...    NaN     21.0          1   
50810        1651            OO       5750  ...    NaN     10.0          1   
67078        1615            XE       2173  ...    NaN      NaN          1   

       CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
5976                  A         0           NaN          NaN      NaN   
15680                 B         0           NaN          NaN      NaN   
16382                 B         0           NaN          NaN      NaN   
20622                 A         0           NaN          NaN      NaN   
22462                 B         0           NaN          NaN      NaN   
30057                 B         0           NaN          NaN      NaN   
33611                 A         0           NaN          NaN      NaN   
37581                 B         0           NaN          NaN      NaN   
40343                 A         0           NaN          NaN      NaN   
50401                 B         0           NaN          NaN      NaN   
50810                 B         0           NaN          NaN      NaN   
67078                 A         0           NaN          NaN      NaN   

       SecurityDelay  LateAircraftDelay  
5976             NaN                NaN  
15680            NaN                NaN  
16382            NaN                NaN  
20622            NaN                NaN  
22462            NaN                NaN  
30057            NaN                NaN  
33611            NaN                NaN  
37581            NaN                NaN  
40343            NaN                NaN  
50401            NaN                NaN  
50810            NaN                NaN  
67078            NaN                NaN  

[12 rows x 29 columns]

Отмененных рейсов, у которых ЕСТЬ информация по DepDelay, 12 штук:

In [41]:
db[(db['Cancelled'] == 1) & (db['DepDelay'].isnull() == False)].shape[0]

12

Всего перелетов, у которых есть информация по DepDelay:

In [42]:
db[db['DepDelay'].isnull() == False].shape[0]

68601

Вывод: число 12 слишком мало против 68601 (все остальные рейсы не были отменены) => можно удалить признак Cancelled, вместе с ним CancellationCode.

Количество пропусков по другим признакам при условии, что информация по DepDelay присутствует:

In [43]:
db[(db['DepDelay'].isnull() == False)].isnull().sum()

Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                157
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  0
ActualElapsedTime      183
CRSElapsedTime           5
AirTime                183
ArrDelay               183
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 157
TaxiOut                  7
Cancelled                0
CancellationCode     68589
Diverted                 0
CarrierDelay         53348
WeatherDelay         53348
NASDelay             53348
SecurityDelay        53348
LateAircraftDelay    53348
dtype: int64

Так как в 53348 строках из 68601 (78%) отсутствует информация о CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, можно исключить и эти признаки.

Исключили из выборки объекты с пропущенным значением целевой переменной и со значением целевой переменной, равным 0, а также исключили признаки Cancelled, CancellationCode, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay. Выделили целевую переменную в отдельный вектор, исключив её из матрицы "объект-признак".

In [44]:
sample = db[(db['DepDelay'].isnull() == False) & (db['DepDelay'] != 0)]
sample = sample.drop(['Cancelled', 'CancellationCode', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'], axis='columns')

sample = sample.reset_index(drop=True)
delay = sample['DepDelay'].values
sample = sample.drop('DepDelay', axis='columns')
sample

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0      2008      6          18          3   2111.0        2055   2308.0   
1      2008      6           4          3   1426.0        1410   1730.0   
2      2008      8           3          7   1143.0        1145   1501.0   
3      2008      1          23          3   1141.0        1144   1323.0   
4      2008      5           4          7    815.0         820   1243.0   
...     ...    ...         ...        ...      ...         ...      ...   
63399  2008      1          26          6    920.0         925   1110.0   
63400  2008      5          12          1   1829.0        1840   2137.0   
63401  2008      5          11          7   1149.0        1125   1336.0   
63402  2008      2          18          1   1906.0        1900   2200.0   
63403  2008     12           6          6    859.0         900    959.0   

       CRSArrTime UniqueCarrier  FlightNum  ... ActualElapsedTime  \
0            2300            WN        242  ...              57.0   
1            1728            XE       2380  ...             124.0   
2            1520            WN       1769  ...             138.0   
3            1335            OO       3802  ...             102.0   
4            1300            WN        399  ...             148.0   
...           ...           ...        ...  ...               ...   
63399        1118            OH       5218  ...             110.0   
63400        2149            DL        794  ...             128.0   
63401        1314            OO       6159  ...              47.0   
63402        2139            NW        641  ...             234.0   
63403        1005            WN        510  ...              60.0   

       CRSElapsedTime  AirTime  ArrDelay  Origin Dest Distance  TaxiIn  \
0                65.0     46.0       8.0     MDW  CLE      307     3.0   
1               138.0    108.0       2.0     IAH  AVL      834     4.0   
2               155.0    125.0     -19.0     MDW  TPA      997     4.0   
3               111.0     79.0     -12.0     SLC  FCA      532     4.0   
4               160.0    136.0     -17.0     LAS  AUS     1090     4.0   
...               ...      ...       ...     ...  ...      ...     ...   
63399           113.0     85.0      -8.0     CVG  BDL      661    12.0   
63400           129.0     86.0     -12.0     ORD  ATL      606    22.0   
63401            49.0     30.0      22.0     ORD  GRR      137     4.0   
63402           219.0    192.0      21.0     RSW  MSP     1416    20.0   
63403            65.0     46.0      -6.0     LAS  BUR      223     2.0   

       TaxiOut  Diverted  
0          8.0         0  
1         12.0         0  
2          9.0         0  
3         19.0         0  
4          8.0         0  
...        ...       ...  
63399     13.0         0  
63400     20.0         0  
63401     13.0         0  
63402     22.0         0  
63403     12.0         0  

[63404 rows x 21 columns]

**13. (0.5 балла)** Обратите внимание, что признаки DepTime, CRSDepTime, ArrTime, CRSArrTime приведены в формате hhmm, в связи с чем будет не вполне корректно рассматривать их как вещественные.

Преобразуйте каждый признак FeatureName из указанных в пару новых признаков FeatureName\_Hour, FeatureName\_Minute, разделив каждое из значений на часы и минуты. Не забудьте при этом исключить исходный признак из выборки. В случае, если значение признака отсутствует, значения двух новых признаков, его заменяющих, также должны отсутствовать. 

Например, признак DepTime необходимо заменить на пару признаков DepTime_Hour, DepTime_Minute. При этом, например, значение 155 исходного признака будет преобразовано в значения 1 и 55 признаков DepTime_Hour, DepTime_Minute соответственно.

In [45]:
def get_hh(time):
    if np.isnan(time):
        return np.nan
    time = int(time)
    if time // 100 < 1:
        return np.nan
    else:
        return int(str(time)[:-2])

def get_mm(time):
    if np.isnan(time):
        return np.nan
    time = int(time)
    if time // 100 < 1:
        return np.nan
    else:
        return int(str(time)[-2:])

In [46]:
for item in ['DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime']:
    item_Hour = item + '_Hour'
    item_Minute = item + '_Minute'
    sample[item_Hour] = np.array(sample[item].apply(get_hh))
    sample[item_Minute] = np.array(sample[item].apply(get_mm))
    sample = sample.drop(item, axis=1)

In [47]:
sample.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'UniqueCarrier',
       'FlightNum', 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime',
       'AirTime', 'ArrDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn',
       'TaxiOut', 'Diverted', 'DepTime_Hour', 'DepTime_Minute',
       'CRSDepTime_Hour', 'CRSDepTime_Minute', 'ArrTime_Hour',
       'ArrTime_Minute', 'CRSArrTime_Hour', 'CRSArrTime_Minute'],
      dtype='object')

In [48]:
sample.dtypes

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Diverted               int64
DepTime_Hour         float64
DepTime_Minute       float64
CRSDepTime_Hour      float64
CRSDepTime_Minute    float64
ArrTime_Hour         float64
ArrTime_Minute       float64
CRSArrTime_Hour      float64
CRSArrTime_Minute    float64
dtype: object

**14. (0.5 балла)** Некоторые из признаков, отличных от целевой переменной, могут оказывать чересчур значимое влияние на прогноз, поскольку по своему смыслу содержат большую долю информации о значении целевой переменной. Изучите описание датасета и исключите признаки, сильно коррелирующие с ответами. Ваш выбор признаков для исключения из выборки обоснуйте. Кроме того, исключите признаки TailNum и Year.

In [ ]:
# Your code here

**15. (1 балл)** Приведем данные к виду, пригодному для обучения линейных моделей. Для этого вещественные признаки надо отмасштабировать, а категориальные — привести к числовому виду. Также надо устранить пропуски в данных.

В первую очередь поймем, зачем необходимо применять масштабирование. Следующие ячейки с кодом построят гистограммы для 3 вещественных признаков выборки.

In [ ]:
X['DepTime_Hour'].hist(bins=20)

In [ ]:
X['TaxiIn'].hist(bins=20)

In [ ]:
X['FlightNum'].hist(bins=20)

Какую проблему вы наблюдаете на этих графиках? Как масштабирование поможет её исправить?

Некоторые из признаков в нашем датасете являются категориальными. Типичным подходом к работе с ними является бинарное, или [one-hot-кодирование](https://en.wikipedia.org/wiki/One-hot).

Реализуйте функцию transform_data, которая принимает на вход DataFrame с признаками и выполняет следующие шаги:
1. Замена пропущенных значений на нули для вещественных признаков и на строки 'nan' для категориальных.
2. Масштабирование вещественных признаков с помощью [StandardScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html).
3. One-hot-кодирование категориальных признаков с помощью [DictVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html) или функции [pd.get_dummies](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html).

Метод должен возвращать преобразованный DataFrame, который должна состоять из масштабированных вещественных признаков и закодированных категориальных (исходные признаки должны быть исключены из выборки).

In [ ]:
def transform_data(data):
    # Your code here

Примените функцию transform_data к данным. Сколько признаков получилось после преобразования?

In [ ]:
# Your code here

**16. (0.5 балла)** Разбейте выборку и вектор целевой переменной на обучение и контроль в отношении 70/30 (для этого можно использовать, например, функцию [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.train_test_split.html)). 

In [ ]:
# Your code here

### Scikit-learn

<img src = "https://pp.vk.me/c4534/u35727827/93547647/x_d31c4463.jpg">
Теперь, когда мы привели данные к пригодному виду, попробуем решить задачу при помощи метода наименьших квадратов. Напомним, что данный метод заключается в оптимизации функционала $MSE$:

$$MSE(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 \to \min_{w},$$

где $\{ (x_i, y_i ) \}_{i=1}^l$ — обучающая выборка, состоящая из $l$ пар объект-ответ.

Заметим, что решение данной задачи уже реализовано в модуле sklearn в виде класса [LinearRegression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression).

**17. (0.5 балла)** Обучите линейную регрессию на 1000 объектах из обучающей выборки и выведите значения $MSE$ и $R^2$ на этой подвыборке и контрольной выборке (итого 4 различных числа). Проинтерпретируйте полученный результат — насколько качественные прогнозы строит полученная модель? Какие проблемы наблюдаются в модели?

**Подсказка**: изучите значения полученных коэффициентов $w$, сохраненных в атрибуте coef_ объекта LinearRegression.

In [ ]:
# Your code here

Для решения описанных вами в предыдущем пункте проблем используем L1- или L2-регуляризацию, тем самым получив Lasso и Ridge регрессии соответственно и изменив оптимизационную задачу одним из следующих образов:
$$MSE_{L1}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_1 \to \min_{w},$$
$$MSE_{L2}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_2^2 \to \min_{w},$$

где $\alpha$ — коэффициент регуляризации. Один из способов его подбора заключается в переборе некоторого количества значений и оценке качества на кросс-валидации для каждого из них, после чего выбирается значение, для которого было получено наилучшее качество.

In [ ]:
# Your code here

__18. (1 балл) __ Обучение линейной регрессии.



Обучите линейную регрессию с $L_1$ (Lasso) и $L_2$ (Ridge) регуляризаторами (используйте параметры по умолчанию). Посмотрите, какое количество коэффициентов близко к 0 (степень близости к 0 определите сами из разумных пределов). Постройте график зависимости числа ненулевых коэффициентов от коэффицента регуляризации (перебирайте значения по логарифмической сетке от $10^{-3}$ до $10^3$). Согласуются ли результаты с вашими ожиданиями?

In [ ]:
# Your code here
# ...

Посчитайте для Ridge-регрессии следующие метрики: $RMSE$, $MAE$, $R^2$.

In [ ]:
# Your code here
# ...

Подберите на обучающей выборке для Ridge-регрессии коэффициент регуляризации (перебирайте значения по логарифмической сетке от $10^{-3}$ до $10^3$) для каждой из метрик при помощи кросс-валидации c 5 фолдами на тех же 1000 объектах. Для этого воспользуйтесь GridSearchCV и KFold из sklearn. Постройте графики зависимости фукнции потерь от коэффициента регуляризации. Посчитайте те же метрики снова. Заметно ли изменилось качество?

Для выполнения данного задания вам могут понадобиться реализованные в библиотеке объекты [LassoCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html), [RidgeCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html) и [KFold](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.KFold.html).

In [ ]:
# Your code here
# ...

__19. (0.5 балла)__ Поиск объектов-выбросов


Как известно, MSE сильно штрафует за большие ошибки на объектах-выбросах. С помощью cross_val_predict сделайте Out-of-Fold предсказания для обучающей выборки. Посчитайте ошибки и посмотрите на их распределение (plt.hist). Что вы видите?